Sources/Datadog/Tracing/DDSpan.swift

```diff
@@ -123,38 +109,42 @@ internal class DDSpan: OTSpan {
     }
 
     func log(fields: [String: Encodable], timestamp: Date) {
-        if warnIfFinished("log(fields:timestamp:)") {
-            return
-        }
-        ddTracer.queue.async {
+        queue.async {
+            if self.warnIfFinished("log(fields:timestamp:)") {
+                return
+            }
             self.unsafeLogFields.append(fields)
         }
         sendSpanLogs(fields: fields, date: timestamp)
     }
 
     func finish(at time: Date) {
-        if warnIfFinished("finish(at:)") {
-            return
+        let isFinished: Bool = queue.sync {
+            let wasFinished = self.warnIfFinished("finish(at:)")
+            self.unsafeIsFinished = true
+            return wasFinished
         }
-        isFinished = true
-        if let activity = activityReference {
-            ddTracer.activeSpansPool.removeSpan(activityReference: activity)
+
+        if !isFinished {
+            if let activity = activityReference {
+                ddTracer.activeSpansPool.removeSpan(activityReference: activity)
+            }
+            sendSpan(finishTime: time)
         }
-        sendSpan(finishTime: time)
     }
 
     // MARK: - Writting SpanEvent
 
     /// Sends span event for given `DDSpan`.
     private func sendSpan(finishTime: Date) {
-        // Baggage items must be read before entering the `tracer.queue` as it uses that queue for internal sync.
+        // Baggage items must accessed outside `tracer.queue` as it uses that queue for internal sync.
```

Comment:
`must be accessed` ?

Reply:
Fixing also in few other places, thx 🙂 👌 
```suggestion
        // Baggage items must be accessed outside the `tracer.queue` as it uses that queue for internal sync.
```